tortoisebeacon/tortoisebeacon.go

```diff
@@ -0,0 +1,652 @@
+				}
+
+				hashesAgainst := make([]types.Hash32, 0, len(timelyProposals))
+				for _, p := range timelyProposals {
+					hashesAgainst = append(hashesAgainst, hashATXList(p))
+				}
+
+				m := NewVotingMessage(epoch, round, hashesFor, hashesAgainst)
+				tb.messageSender.Send(m)
+			} else {
+				// next rounds, send vote
+				// construct a message that points to all messages from previous round received by δ
+				votesFor, votesAgainst := tb.calculateVotes(epoch, round)
+
+				m := NewVotingMessage(epoch, round, votesFor, votesAgainst)
+				tb.messageSender.Send(m)
+			}
+
+		case <-tb.CloseChannel():
+			return
+		}
+	}
+}
+
+// TODO(nkryuchkov): send and handle votes diff instead of votes list
+// TODO(nkryuchkov): refactor this
+func (tb *TortoiseBeacon) calculateVotes(epoch types.EpochID, round int) (votesFor []types.Hash32, votesAgainst []types.Hash32) {
+	votesForMap := make(map[types.Hash32]int)
+	votesAgainstMap := make(map[types.Hash32]int)
+
+	tb.timelyProposalsMu.Lock()
+
+	for _, proposals := range tb.timelyProposalsList[epoch] {
+		proposalsHash := hashATXList(proposals)
+
+		countFor := 1
+		countAgainst := 0
+		for i := 1; i < round; i++ {
+			key := EpochRoundPair{
+				EpochID: epoch,
+				Round:   i, // proposals from the previous round are needed
+			}
+
+			// TODO(nkryuchkov): use map
+			for _, vf := range tb.votesFor[key] {
+				if vf == hashATXList(proposals) {
+					countFor++
+				}
+			}
+
+			// TODO(nkryuchkov): use map
+			for _, vf := range tb.votesAgainst[key] {
+				if vf == proposalsHash {
+					countAgainst++
+				}
+			}
+		}
+
+		votesForMap[proposalsHash] = countFor
+		votesAgainstMap[proposalsHash] = countAgainst
+	}
+
+	for _, proposals := range tb.delayedProposalsList[epoch] {
+		proposalsHash := hashATXList(proposals)
+
+		countFor := 0
+		countAgainst := 1
+		for i := 1; i < round; i++ {
+			key := EpochRoundPair{
+				EpochID: epoch,
+				Round:   i, // proposals from the previous round are needed
+			}
+
+			// TODO(nkryuchkov): use map
+			for _, vf := range tb.votesFor[key] {
+				if vf == hashATXList(proposals) {
+					countFor++
+				}
+			}
+
+			// TODO(nkryuchkov): use map
+			for _, vf := range tb.votesAgainst[key] {
+				if vf == proposalsHash {
+					countAgainst++
+				}
+			}
+		}
+
+		if _, ok := votesForMap[proposalsHash]; !ok {
+			votesForMap[proposalsHash] = countFor
+		} else {
+			votesForMap[proposalsHash] += countFor
+		}
+
+		if _, ok := votesAgainstMap[proposalsHash]; !ok {
+			votesAgainstMap[proposalsHash] = countAgainst
+		} else {
+			votesAgainstMap[proposalsHash] += countAgainst
+		}
+	}
+
+	tb.timelyProposalsMu.Unlock()
+
+	votesFor = make([]types.Hash32, 0)
+	votesAgainst = make([]types.Hash32, 0)
+
+	for k, v := range votesForMap {
+		if v > threshold() || tb.WeakCoin(epoch, round) {
+			votesFor = append(votesFor, k)
+		}
+	}
+
+	for k, v := range votesAgainstMap {
+		if v > threshold() || !tb.WeakCoin(epoch, round) {
```

Comment:
theoretically you need to vote against messages that came in late - but as said before I think we can drop this part